src/main.cpp

```diff
@@ -0,0 +1,1616 @@
+            &swapchain_image_count,
+            swapchain_images.data());
+
+    swapchain_image_views = create_image_views(
+            logical_device,
+            swapchain_images,
+            swapchain_image_format);
+
+    render_pass = create_render_pass(
+            logical_device,
+            swapchain_image_format);
+
+    pipeline_layout = create_pipeline_layout(logical_device);
+
+    Log::i("Creating Vulkan graphics pipeline");
+    graphics_pipeline = create_graphics_pipeline(
+            logical_device,
+            swapchain_extent,
+            pipeline_layout,
+            render_pass);
+
+    swapchain_framebuffers = create_framebuffers(
+            logical_device,
+            swapchain_image_views,
+            render_pass,
+            swapchain_extent);
+
+    command_pool = create_command_pool(
+            logical_device,
+            family_indices);
+
+    command_buffer = create_command_buffer(
+            logical_device,
+            command_pool);
+
+    semaphore_image_available = create_vulkan_semaphore(logical_device);
+    semaphore_render_finished = create_vulkan_semaphore(logical_device);
+    fence_in_flight = create_vulkan_fence(logical_device);
+
+    draw_frame(
+            logical_device,
+            fence_in_flight,
+            swapchain,
+            semaphore_image_available,
+            semaphore_render_finished,
+            command_buffer,
+            render_pass,
+            swapchain_framebuffers,
+            swapchain_extent,
+            graphics_pipeline,
+            drawing_queue,
+            presentation_queue);
+
+#if 0
+    // TODO Main loop here
+    while (1 /* loop until window close event */) {
+    }
+#endif
+
+    Log::i("Sleeping for 1 second");
+    std::this_thread::sleep_for(
+            std::chrono::milliseconds(1000));
+
+    // SDL createMainRenderer # <- necessary here?
+
+    // SDL setRenderDrawColor # <- necessary here?
+
+    msg_temp = g_program_name;
+    msg_temp += " shutting down";
+    Log::i(msg_temp);
+
+    vkDestroySemaphore(logical_device, semaphore_image_available, nullptr);
+    vkDestroySemaphore(logical_device, semaphore_render_finished, nullptr);
+    vkDestroyFence(logical_device, fence_in_flight, nullptr);
+
+    vkDestroyCommandPool(logical_device, command_pool, nullptr);
+
+    for (VkFramebuffer framebuffer : swapchain_framebuffers) {
+        vkDestroyFramebuffer(logical_device, framebuffer, nullptr);
+    }
+
+    vkDestroyPipeline(logical_device, graphics_pipeline, nullptr);
+    vkDestroyPipelineLayout(logical_device, pipeline_layout, nullptr);
+    vkDestroyRenderPass(logical_device, render_pass, nullptr);
+
+    for (VkImageView image_view : swapchain_image_views) {
+        vkDestroyImageView(logical_device, image_view, nullptr);
+    }
+
+    vkDestroySwapchainKHR(logical_device, swapchain, nullptr);
+    vkDestroySurfaceKHR(vulkan_instance, surface, nullptr);
+    vkDestroyDevice(logical_device, nullptr);
```

Comment:
Surface is created before logical device, so please remove logical device before surface.

Reply:
Order changed in: https://github.com/pottumuusi/neotetris/pull/1/commits/0596f64d4ae5c59ab4c7ea7dc1d48778759d376c